adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -744,41 +916,215 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+          .flatMap(f => f.filter(_.isDefined).map(g => g.get))
+          .map(f => ((f._2._1, f._1), f._2))
+          .repartitionAndSortWithinPartitions(
+            new ReferenceRegionRangePartitioner(numPartitions))
+          .map(_._2)
+      }
+    // here we get the new partition map
+    // for now, we can't use the left RDD's partition map because
+    // the bounds don't always include the values we copartitioned with.
+    // in other words, joins allow data that starts before a given
+    // ReferenceRegion to combine given they overlap.
+    // this means that we cannot use the bounds of the left RDD because
+    // they may not necessarily be the bounds of the right RDD
+    val newPartitionMap = finalPartitionedRDD.mapPartitions(iter => {
+      getRegionBoundsFromPartition(iter)
+    }, preservesPartitioning = true).collect
+
+    replaceRdd(finalPartitionedRDD.values, Some(newPartitionMap))
+  }
+
+  /**
+   * Gets the partition bounds from a ReferenceRegion keyed Iterator
+   *
+   * @param iter The data on a given partition. ReferenceRegion keyed
+   * @return The bounds of the ReferenceRegions on that partition, in an Iterator
+   */
+  private def getRegionBoundsFromPartition(iter: Iterator[(ReferenceRegion, T)]): Iterator[Option[(ReferenceRegion, ReferenceRegion)]] = {
+
+    if (iter.isEmpty) {
+      // This essentially means that there is no data on the partition, so we
+      // have no bounds
+      Iterator(None)
+    } else {
+      val firstRegion = iter.next._1
+      if (iter.hasNext) {
+        var lastRegion = iter.next._1
+        while (iter.hasNext) lastRegion = iter.next._1
+        Iterator(Some((firstRegion, lastRegion)))
+        //only one ReferenceRegion on this partition
+      } else Iterator(Some((firstRegion, firstRegion)))
+    }
+  }
+
+  /**
+   * Sort and repartition the data. We use Sparks sortByKey on the
+   * ReferenceRegions of a given record. This gets us data that is
+   * partitioned to be roughly uniform in size across all nodes.
+   *
+   * @param partitions The number of partitions to split the data into
+   * @return A SortedGenomicRDDMixIn that contains the sorted and partitioned
+   *         RDD
+   */
+  def repartitionAndSort(partitions: Int = rdd.partitions.length)(implicit c: ClassTag[T]): U = {
+    // here we let spark do the heavy lifting
+    val partitionedRDD = flattenRddByRegions()
+      .sortBy(f => f._1, ascending = true, partitions)
+    replaceRdd(partitionedRDD.values,
+      Some(partitionedRDD.mapPartitions(iter =>
+        getRegionBoundsFromPartition(iter), preservesPartitioning = true).collect))
   }
 }
 
 private case class GenericGenomicRDD[T](rdd: RDD[T],
                                         sequences: SequenceDictionary,
-                                        regionFn: T => Seq[ReferenceRegion]) extends GenomicRDD[T, GenericGenomicRDD[T]] {
+                                        regionFn: T => Seq[ReferenceRegion],
```

Comment:
Small nit. These lines are starting to get very long. Can we break `[T](rdd: RDD[T],` with a newline before `rdd`? This will lead to better indentation/shorter lines.